sky/serve/load_balancer.py

```diff
@@ -70,30 +74,110 @@ def _sync_with_controller(self):
                         'ready_replica_urls')
                 except requests.RequestException as e:
-                    print(f'An error occurred: {e}')
+                    logger.error(f'An error occurred: {e}')
                 else:
                     logger.info(f'Available Replica URLs: {ready_replica_urls}')
-                    self._load_balancing_policy.set_ready_replicas(
-                        ready_replica_urls)
+                    with self._lock:
+                        self._load_balancing_policy.set_ready_replicas(
+                            ready_replica_urls)
+                        for replica_url in ready_replica_urls:
+                            if replica_url not in self._client_pool:
+                                # TODO(tian): Support HTTPS.
+                                self._client_pool[
+                                    replica_url] = httpx.AsyncClient(
+                                        base_url=f'http://{replica_url}')
+                        closed_urls = []
+                        for replica_url, client in self._client_pool.items():
+                            if replica_url not in ready_replica_urls:
+                                asyncio.run(client.aclose())
+                                closed_urls.append(replica_url)
+                        for replica_url in closed_urls:
+                            del self._client_pool[replica_url]
             time.sleep(constants.LB_CONTROLLER_SYNC_INTERVAL_SECONDS)
 
-    async def _redirect_handler(self, request: fastapi.Request):
-        self._request_aggregator.add(request)
-        ready_replica_url = self._load_balancing_policy.select_replica(request)
-
-        if ready_replica_url is None:
-            raise fastapi.HTTPException(status_code=503,
-                                        detail='No ready replicas. '
-                                        'Use "sky serve status [SERVICE_NAME]" '
-                                        'to check the replica status.')
+    async def _proxy_request_to(
+        self, url: str, request: fastapi.Request
+    ) -> Union[fastapi.responses.Response, Exception]:
+        """Proxy the request to the specified URL.
 
-        path = f'http://{ready_replica_url}{request.url.path}'
-        logger.info(f'Redirecting request to {path}')
-        return fastapi.responses.RedirectResponse(url=path)
+        Returns:
+            The response from the endpoint replica. Return the exception
+            encountered if anything goes wrong.
+        """
+        logger.info(f'Proxy request to {url}')
+        try:
+            # We defer the get of the client here on purpose, for case when the
+            # replica is ready in `_proxy_with_retries` but refreshed before
+            # entering this function. In that case we will return an error here
+            # and retry to find next ready replica.
+            client = self._client_pool.get(url, None)
```

Comment:
Do we need to lock the `get` here? I am confused about what the lock is used for.

Reply:
Good catch! Yeah we need to lock here.. It is mainly to avoid updating and getting the ready replcias & clients at the same time. Fixed now!